[Week3] Spring Web으로 ToDo REST API 만들기-테스트

JIY0UNG opened the pull request:
안녕하세요😊 드디어 테스트 코드를 작성해 보게 되었네요!!
오늘은 HelloController에 대한 유닛 테스트와 MockMvc 테스트, 그리고 Task 모델의 getId메서드에 대한 유닛 테스트 코드를 작성했습니다.

다른 분들 PR을 둘러보다가 [https://johngrib.github.io/wiki/junit5-nested/](https://johngrib.github.io/wiki/junit5-nested/) 이 글을 참조해 보라고 하신 리뷰를 보고, 저도 계층 구조로 만들어 보았습니다. 맞게 잘 한 건지는 모르겠네요..

리뷰 부탁드립니다~!

JIY0UNG committed:
2주 차 과제와 같은 상태로 만들어라

2주 차 과제 코드에 이어서 테스트를 만들기 위해 일부 수정하였습니다.

JIY0UNG committed:
DTO를 통해서 Task에 대한 요청 및 응답을 하도록 변경하라

Task 클래스는 DB에 직접적으로 접근하는 클래스인데, 이를 그대로 외부에 노출시키면 예기치 못한 변경이 일어날 수 있다는 위험이 있습니다. 따라서 사용자와 데이터를 주고받을 때 DTO를 통해 주고받도록 수정했습니다.

JIY0UNG committed:
HelloController를 만들어라

JIY0UNG committed:
HelloController에 대한 유닛 테스트를 만들어라

JIY0UNG committed:
HelloController에 대한 MockMVC 테스트를 만들어라

JIY0UNG committed:
title유효성 검사를 Task 클래스에서 하도록 수정하라

- 기존에는 요청으로 들어온 title에 대한 유효성 검사를 DTO에서 했었으나, DTO의 역할은 '데이터 전달'이기 때문에 그 외의 일은 하지 않는 것이 좋다는 생각이 들었습니다.
그래서 Task 클래스에서 title에 대한 유효성 검사를 하도록 수정했습니다.
- 유효성 검사를 먼저 한 후에 id를 부여하기 위해서 Task 클래스에 setId메서드를 만들었습니다.

JIY0UNG committed:
Task 모델의 getId메서드에 대한 유닛 테스트 코드를 작성하라

JIY0UNG committed:
HelloController 유닛 테스트를 계층구조로 수정하라

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,30 @@
+        class 만약_id가_1인_Task_객체가_주어진다면 {
+            private Task task;
+
+            @BeforeEach
+            void setUp() {
+                task = new Task("과제하기");
+                task.setId(1L);
+            }
```
이것이 `만약_id가_1인_Task_객체가_주어진다면` 을 말하는 것일겁니다! 

아래 링크를 보고 한번 수용해보시면 좋은 경험이 되실거에요!
https://johngrib.github.io/wiki/junit5-nested/#subject-%EB%A9%94%EC%86%8C%EB%93%9C%EC%9D%98-%EC%82%AC%EC%9A%A9

JIY0UNG replied:
넵 해당 부분을 아래와 같은 코드로 캡슐화했습니다!
```java
Task task() {
	Task task = new Task("과제하기");
	task.setId(1L);
	return task;
}
```

JIY0UNG committed:
Task 모델의 getId테스트에서 테스트 대상을 실행하는 코드를 캡슐화하라

테스트 대상에 대한 코드를 task()로 캡슐화하여, 테스트 대상에 변경이 발생하더라도 이에 딸린 테스트 코드들을 모두 수정할 필요가 없도록 했습니다.

JIY0UNG committed:
Task 모델의 메서드들에 대한 유닛 테스트 코드를 작성하라

Task 모델의 생성자, setId, getTitle, updateTitle, validateTitle 메서드에 대한 테스트 코드를 작성했습니다.

JIY0UNG committed:
TaskService의 getTasks 메서드에 대한 유닛 테스트 코드를 작성하라

JIY0UNG commented:
Task 모델의 나머지 메서드들과 TaskService의 getTasks 메서드에 대한 유닛 테스트 코드를 작성해 보았습니다~
리뷰 부탁드립니다!!

JIY0UNG committed:
TaskRequest 클래스에 title 매개변수를 가지는 생성자를 추가하라

TaskService 테스트에서 미리 Task를 추가하는 작업을 할 때 특정 title을 가지는 TaskRequest 객체를 생성해 줘야 합니다.
객체의 일관성을 유지하기 힘든 setter보다 생성자를 통해서 title을 설정해 주는 것이 더 좋다고 생각하여 해당 생성자를 추가하였습니다.

JIY0UNG committed:
TaskService에 대한 유닛 테스트 코드를 작성하라

JIY0UNG committed:
TaskController에 대한 유닛 테스트 코드를 작성하라

JIY0UNG commented:
안녕하세요!
오늘은 TaskService와 TaskController에 대한 유닛 테스트 코드를 모두 작성했습니다.

어제 PR 올렸던 TaskService의 getTasks에 대한 테스트 코드는 조건에 따른 결과를 알 수 있도록 좀 더 구체화하는 것이 좋을 것 같아서 수정했습니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,294 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@DisplayName("TaskServiceTest 클래스")
```
```suggestion
@DisplayName("TaskServiceTest 클래스")
@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
```

이렇게 하고 아래있는 어노테이션은 전부 지워도 될 것 같아요!

JIY0UNG replied:
`@Nested`처럼 클래스마다 써줘야 하는 건 줄 알았는데 이렇게 해도 되는 거였군요!
불필요한 코드를 줄일 수 있어서 좋네요👍👍

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -11,41 +12,41 @@
 @CrossOrigin
 public class TaskController {
-    private TaskService taskService;
+    private final TaskService taskService;
```
좋아요! 테스트 뿐만 아니라 기존 코드도 개선해보는군요! 그렇다면 저도 기존코드에 대한 욕심을 함께 부려보겠습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/dto/TaskRequest.java:
```diff
@@ -0,0 +1,15 @@
+public class TaskRequest {
+    private String title;
+
+    public TaskRequest() {}
```
저도 알아요! 이것이 어쩔 수 없다는 것을 ㅠㅜㅠㅜ 
하지만 우리는 기본생성자를 없앴을 때, `TaskRequest` 가 더 빛날 거라는 것을 알 고 있어요!

한번 어떻게 없앨 수 있을지 고민해보아요!

JIY0UNG committed:
@DisplayNameGeneration을 최상위 클래스에 선언하는 것으로 수정하라

기존에 계층별로 나눈 각 클래스들의 이름을 한글로 짓기 위해 클래스마다 `@DisplayNameGeneration` 어노테이션을 붙여줬었는데,
어차피 모든 내부 클래스에 적용되는 공통 어노테이션이므로 최상위 클래스에 선언함으로써 클래스마다 어노테이션을 달지 않아도 되게 수정했습니다.

JIY0UNG committed:
테스트 메서드의 DisplayName 어노테이션을 삭제하라

최상위 클래스에 @DisplayNameGeneration가 선언되어 있어서 @DisplayName의 기능을 이미 수행하고 있습니다. 따라서 @DisplayName이 굳이 필요하지 않다고 생각해서 삭제했습니다.

JIY0UNG committed:
TaskController에 대한 MockMvc 테스트 코드를 작성하라

JIY0UNG commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,260 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@DisplayName("TaskControllerTest 클래스")
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+class TaskControllerTest {
+    @Nested
+    class getTasks_메서드는 {
+        @Nested
+        class 만약_Task가_추가되지_않은_상태라면 {
+            TaskController taskController() {
+                return new TaskController(new TaskService());
+            }
+
+            @Test
+            void 크기가_0인_TaskResponse_리스트를_반환한다() {
```
`@DisplayNameGeneration`과 `@DisplayName`에 대해서 좀 찾아보던 중
> 클래스나 메서드명을 아예 한글로 짓는 것은 네이밍의 일관성이 깨질 수도 있다(ex. User → 유저, 회원, 사용자 등)

라는 의견이 담긴 글을 읽게 되었는데요, 이 말을 보니까 그럴 수도 있겠다 싶더라구요.

그래서 기존의 클래스와 메서드명을 모두 영어로 변경 후 `@DisplayName`으로 한글 설명을 넣는 것을 고려하고 있는데, 혹시 어떻게 생각하시나요??

JIY0UNG commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -0,0 +1,338 @@
+            void setUp() {
+                Long id = 1L;
+                given(taskService.updateTask(id, new TaskRequest(""))).willThrow(new NotFoundException(id));
+            }
+
+            @Test
+            void status_code_400을_응답한다() throws Exception {
+                ObjectMapper objectMapper = new ObjectMapper();
+                String jsonRequest = objectMapper.writeValueAsString(new TaskRequest(""));
+
+                mockMvc.perform(put("/tasks")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(jsonRequest))
+                        .andExpect(status().isBadRequest())
+                        .andExpect(result -> assertTrue(result.getResolvedException() instanceof BadRequestException));
+            }
+        }
+
+        @Nested
+        class 만약_유효한_id와_title이_주어진다면 {
+            @BeforeEach
+            void setUp() {
+                Long id = 1L;
+                given(taskService.updateTask(id, new TaskRequest("춤추기"))).willThrow(new NotFoundException(id));
+            }
+
+            @Test
+            void status_code_200을_응답한다() throws Exception {
+                ObjectMapper objectMapper = new ObjectMapper();
+                String jsonRequest = objectMapper.writeValueAsString(new TaskRequest("춤추기"));
+
+                mockMvc.perform(put("/tasks/1")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(jsonRequest))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$.id", is(1)))
+                        .andExpect(jsonPath("$.title", is("춤추기")));
+            }
+        }
+    }
```
현재 create, update, patch에 대한 테스트를 실행했을 때 응답이 전부 null로 돌아와서 테스트에 실패하는데, 아무리 찾아봐도 이유를 모르겠습니다ㅠㅠ

아래 코드와 같이 content에 직접 json형태의 String을 적어보기도 했는데 똑같이 null을 응답해서 실패하더라구요..
```java
mockMvc.perform(post("/tasks")
                .contentType(MediaType.APPLICATION_JSON)
                .content("{\"title\":\"과제하기\"}"))
...
```

'create메서드-유효한 title이 주어졌을 경우'를 실행했을 때 출력된 요청,응답 메시지는 다음과 같습니다.
```bash
MockHttpServletRequest:
      HTTP Method = POST
      Request URI = /tasks
       Parameters = {}
          Headers = [Content-Type:"application/json;charset=UTF-8", Content-Length:"24"]
             Body = {"title":"과제하기"}
    Session Attrs = {}

Handler:
             Type = com.codesoom.assignment.controllers.TaskController
           Method = com.codesoom.assignment.controllers.TaskController#create(TaskRequest)

Async:
    Async started = false
     Async result = null

Resolved Exception:
             Type = null

ModelAndView:
        View name = null
             View = null
            Model = null

FlashMap:
       Attributes = null

MockHttpServletResponse:
           Status = 201
    Error message = null
          Headers = [Vary:"Origin", "Access-Control-Request-Method", "Access-Control-Request-Headers"]
     Content type = null
             Body = 
    Forwarded URL = null
   Redirected URL = null
          Cookies = []
```

ssisksl77 replied:
`TaskRequest` 에 equals 를 추가해보세요! 그것이 시작이 될 거에요!

JIY0UNG reviewed (commented):
컨트롤러에 대한 MockMvc 테스트를 모두 작성했으나, 알 수 없는 오류로 인해 create, update, patch 메서드에 대한 테스트는 모두 실패하고 있는 상태네요😥 리뷰 부탁드립니다!

ssisksl77 commented:
커밋의 제목에서 `@`는 어노테이션이 아닌 다른 용도로 사용되어요! 바로 깃헙관련링크로 변환됩니다!
